File: apps/mobile/src/lib/centipedes.ts
/**
 * Centipede API client — ADR-005 Phase C3.
 *
 * Covers the C2 backend surface so mobile screens have a single
 * import path for everything centipede-related:
 *
 *  - Centipede CRUD                       /centipedes/
 *  - CentipedeSpecies catalog (public)    /centipede-species/
 *  - Per-centipede logs                   /centipedes/{id}/{feedings|molts|substrate-changes|photos}
 *  - QR upload sessions                   /centipedes/{id}/upload-session
 *
 * Centipedes launch on the unified `inverts` table — there's no
 * legacy centipedes table. The /centipedes/ routes are a per-taxon
 * facade over `inverts WHERE taxon='centipede'`. From the mobile
 * client's perspective they behave exactly like /scorpions/.
 *
 * Notably absent: centipedes are SOLITARY (cannibalistic), so there's
 * no /centipede-colonies/ surface here.
 *
 * IMPORTANT — apiClient baseURL already includes `/api/v1` (see the
 * mobile_apiclient_baseurl_double_prefix memory). All paths start at
 * the resource, e.g. '/centipedes/', NOT '/api/v1/centipedes/'.
 *
 * Mirrors the shape of `./scorpions.ts`.
 */
import { apiClient } from '../services/api';

// ---------------------------------------------------------------------------
// Enum-ish types — lowercase wire values, same as the scorpion + tarantula
// APIs. (DB `sex` / `source` are UPPERCASE PG enums but the API surfaces
// them lowercase.)
// ---------------------------------------------------------------------------

export type Sex = 'male' | 'female' | 'unknown';
export type Source = 'bred' | 'bought' | 'wild_caught';
export type Visibility = 'private' | 'public';
export type CentipedeEnclosureType = 'terrestrial' | 'arboreal' | 'fossorial';

export type CareLevel = 'beginner' | 'intermediate' | 'advanced';
export type VenomSeverity = 'mild' | 'moderate' | 'medically_significant';
/** Anamorphic centipedes add segments with each molt; epimorphic species
 * hatch with their full adult segment count. Every species in the v1
 * seed is epimorphic but the schema supports both. */
export type DevelopmentalClass = 'anamorphic' | 'epimorphic';
export type Burrowing = 'none' | 'light' | 'heavy';

export const CARE_LEVEL_LABELS: Record<CareLevel, string> = {
  beginner: 'Beginner',
  intermediate: 'Intermediate',
  advanced: 'Advanced',
};

export const VENOM_SEVERITY_LABELS: Record<VenomSeverity, string> = {
  mild: 'Mild',
  moderate: 'Moderate',
  medically_significant: 'Medically significant',
};

export const DEVELOPMENTAL_CLASS_LABELS: Record<DevelopmentalClass, string> = {
  anamorphic: 'Anamorphic (adds segments with each molt)',
  epimorphic: 'Epimorphic (hatches with adult segment count)',
};

// ---------------------------------------------------------------------------
// Domain types — mirror apps/api/app/schemas/centipede.py + invert_species.py.
// ---------------------------------------------------------------------------

export interface Centipede {
  id: string;
  user_id: string;
  /** Always 'centipede' for rows reached through this lib. */
  taxon: 'centipede';
  species_id: string | null;
  enclosure_id: string | null;

  name: string | null;
  common_name: string | null;
  scientific_name: string | null;
  sex: Sex | null;

  date_acquired: string | null;
  source: Source | null;
  price_paid: string | null;

  /** Centipede + scorpion shared — instar number. Most pet trade
   * Scolopendra fall in the 1-10 range. */
  current_instar: number | null;
  /** Total body length in mm (NOT leg span). */
  current_length_mm: string | null;
  /** Centipede-specific — most Scolopendromorpha have 21. */
  current_segment_count: number | null;
  /** Centipede-specific — usually equals segment count. */
  current_leg_pair_count: number | null;

  enclosure_type: CentipedeEnclosureType | null;
  enclosure_size: string | null;
  substrate_type: string | null;
  substrate_depth: string | null;
  last_substrate_change: string | null;
  target_temp_min: string | null;
  target_temp_max: string | null;
  target_humidity_min: string | null;
  target_humidity_max: string | null;
  water_dish: boolean;
  misting_schedule: string | null;
  last_enclosure_cleaning: string | null;
  enclosure_notes: string | null;

  feeding_paused_reason: string | null;
  feeding_paused_until: string | null;

  photo_url: string | null;
  is_public: boolean;
  visibility: Visibility | null;
  notes: string | null;

  created_at: string;
  updated_at: string | null;
}

export interface CentipedeCreate {
  name?: string | null;
  common_name?: string | null;
  scientific_name?: string | null;
  sex?: Sex | null;
  species_id?: string | null;
  enclosure_id?: string | null;
  date_acquired?: string | null;
  source?: Source | null;
  price_paid?: string | number | null;
  current_instar?: number | null;
  current_length_mm?: string | number | null;
  current_segment_count?: number | null;
  current_leg_pair_count?: number | null;
  enclosure_type?: CentipedeEnclosureType | null;
  enclosure_size?: string | null;
  substrate_type?: string | null;
  substrate_depth?: string | null;
  last_substrate_change?: string | null;
  target_temp_min?: string | number | null;
  target_temp_max?: string | number | null;
  target_humidity_min?: string | number | null;
  target_humidity_max?: string | number | null;
  water_dish?: boolean | null;
  misting_schedule?: string | null;
  last_enclosure_cleaning?: string | null;
  enclosure_notes?: string | null;
  feeding_paused_reason?: string | null;
  feeding_paused_until?: string | null;
  photo_url?: string | null;
  is_public?: boolean;
  notes?: string | null;
  visibility?: Visibility | null;
}

export type CentipedeUpdate = Partial<CentipedeCreate>;

export interface CentipedeSpecies {
  id: string;
  /** Always 'centipede' here — the underlying table is invert_species
   * but the per-taxon route filters. */
  taxon: 'centipede';
  scientific_name: string;
  scientific_name_lower: string;
  slug: string;
  common_names: string[];
  genus: string | null;
  family: string | null;
  order_name: string | null;

  care_level: CareLevel | null;
  temperament: string | null;
  native_region: string | null;
  adult_size: string | null;
  adult_length_min_mm: string | null;
  adult_length_max_mm: string | null;
  growth_rate: string | null;
  type: string | null;

  temperature_min: number | null;
  temperature_max: number | null;
  humidity_min: number | null;
  humidity_max: number | null;

  enclosure_size_juvenile: string | null;
  enclosure_size_adult: string | null;
  substrate_depth: string | null;
  substrate_type: string | null;

  prey_size: string | null;
  feeding_frequency_juvenile: string | null;
  feeding_frequency_adult: string | null;

  water_dish_required: boolean;
  burrowing: Burrowing | null;
  communal_suitable: boolean;

  venom_severity: VenomSeverity | null;
  venom_notes: string | null;

  /** Centipede-specific — drives the care-sheet developmental-biology callout. */
  developmental_class: DevelopmentalClass | null;
  typical_segment_count: number | null;
  typical_leg_pair_count: number | null;

  care_guide: string | null;
  image_url: string | null;

  is_verified: boolean;
  times_kept: number;
  created_at: string;
  updated_at: string | null;
}

// ---------------------------------------------------------------------------
// Log types — same DB tables as tarantula/scorpion logs, but the row is
// parented only by invert_id. The mobile lib exposes them as
// CentipedeFeedingLog / CentipedeMoltLog / CentipedeSubstrateChange so a
// screen can type its state without referencing the polymorphic shape.
// ---------------------------------------------------------------------------

export interface CentipedeFeedingLog {
  id: string;
  invert_id: string | null;
  fed_at: string;
  food_type: string | null;
  food_size: string | null;
  quantity: number;
  accepted: boolean;
  notes: string | null;
  created_at: string;
}

export interface CentipedeFeedingCreate {
  fed_at: string;
  food_type?: string | null;
  food_size?: string | null;
  quantity?: number;
  accepted?: boolean;
  notes?: string | null;
}

export interface CentipedeMoltLog {
  id: string;
  invert_id: string | null;
  molted_at: string;
  premolt_started_at: string | null;
  /** Tarantula-domain field, usually null for centipedes. Reserved
   * for keepers who want to log total body length pre/post-molt via
   * the existing molt schema. */
  leg_span_before: string | null;
  leg_span_after: string | null;
  /** Grams */
  weight_before: string | null;
  weight_after: string | null;
  notes: string | null;
  image_url: string | null;
  created_at: string;
}

export interface CentipedeMoltCreate {
  molted_at: string;
  premolt_started_at?: string | null;
  leg_span_before?: string | number | null;
  leg_span_after?: string | number | null;
  weight_before?: string | number | null;
  weight_after?: string | number | null;
  notes?: string | null;
  image_url?: string | null;
}

export interface CentipedeSubstrateChange {
  id: string;
  invert_id: string | null;
  changed_at: string;
  substrate_type: string | null;
  substrate_depth: string | null;
  reason: string | null;
  notes: string | null;
  created_at: string;
}

export interface CentipedeSubstrateChangeCreate {
  changed_at: string;
  substrate_type?: string | null;
  substrate_depth?: string | null;
  reason?: string | null;
  notes?: string | null;
}

export interface CentipedePhoto {
  id: string;
  url: string;
  thumbnail_url: string | null;
  caption: string | null;
  taken_at: string | null;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Centipede CRUD
// ---------------------------------------------------------------------------

export async function listCentipedes(): Promise<Centipede[]> {
  const { data } = await apiClient.get<Centipede[]>('/centipedes/');
  return data;
}

export async function getCentipede(id: string): Promise<Centipede> {
  const { data } = await apiClient.get<Centipede>(`/centipedes/${id}`);
  return data;
}

export async function createCentipede(
  payload: CentipedeCreate,
): Promise<Centipede> {
  const { data } = await apiClient.post<Centipede>('/centipedes/', payload);
  return data;
}

export async function updateCentipede(
  id: string,
  payload: CentipedeUpdate,
): Promise<Centipede> {
  const { data } = await apiClient.put<Centipede>(`/centipedes/${id}`, payload);
  return data;
}

export async function deleteCentipede(id: string): Promise<void> {
  await apiClient.delete(`/centipedes/${id}`);
}

// ---------------------------------------------------------------------------
// Centipede species catalog (public reads)
// ---------------------------------------------------------------------------

export async function listCentipedeSpecies(options?: {
  skip?: number;
  limit?: number;
  verifiedOnly?: boolean;
  careLevel?: CareLevel;
}): Promise<CentipedeSpecies[]> {
  const params: Record<string, string | number | boolean> = {};
  if (options?.skip != null) params.skip = options.skip;
  if (options?.limit != null) params.limit = options.limit;
  if (options?.verifiedOnly) params.verified_only = true;
  if (options?.careLevel) params.care_level = options.careLevel;
  const { data } = await apiClient.get<CentipedeSpecies[]>(
    '/centipede-species/',
    { params },
  );
  return data;
}

export async function searchCentipedeSpecies(
  q: string,
  limit = 10,
): Promise<CentipedeSpecies[]> {
  const { data } = await apiClient.get<CentipedeSpecies[]>(
    '/centipede-species/search',
    { params: { q, limit } },
  );
  return data;
}

export async function getCentipedeSpecies(id: string): Promise<CentipedeSpecies> {
  const { data } = await apiClient.get<CentipedeSpecies>(
    `/centipede-species/${id}`,
  );
  return data;
}

export async function getCentipedeSpeciesBySlug(
  slug: string,
): Promise<CentipedeSpecies> {
  const { data } = await apiClient.get<CentipedeSpecies>(
    `/centipede-species/by-slug/${slug}`,
  );
  return data;
}

// ---------------------------------------------------------------------------
// Per-centipede logs
// ---------------------------------------------------------------------------

export async function listCentipedeFeedings(
  centipedeId: string,
): Promise<CentipedeFeedingLog[]> {
  const { data } = await apiClient.get<CentipedeFeedingLog[]>(
    `/centipedes/${centipedeId}/feedings`,
  );
  return data;
}

export async function createCentipedeFeeding(
  centipedeId: string,
  payload: CentipedeFeedingCreate,
): Promise<CentipedeFeedingLog> {
  const { data } = await apiClient.post<CentipedeFeedingLog>(
    `/centipedes/${centipedeId}/feedings`,
    payload,
  );
  return data;
}

export async function listCentipedeMolts(
  centipedeId: string,
): Promise<CentipedeMoltLog[]> {
  const { data } = await apiClient.get<CentipedeMoltLog[]>(
    `/centipedes/${centipedeId}/molts`,
  );
  return data;
}

export async function createCentipedeMolt(
  centipedeId: string,
  payload: CentipedeMoltCreate,
): Promise<CentipedeMoltLog> {
  const { data } = await apiClient.post<CentipedeMoltLog>(
    `/centipedes/${centipedeId}/molts`,
    payload,
  );
  return data;
}

export async function listCentipedeSubstrateChanges(
  centipedeId: string,
): Promise<CentipedeSubstrateChange[]> {
  const { data } = await apiClient.get<CentipedeSubstrateChange[]>(
    `/centipedes/${centipedeId}/substrate-changes`,
  );
  return data;
}

export async function createCentipedeSubstrateChange(
  centipedeId: string,
  payload: CentipedeSubstrateChangeCreate,
): Promise<CentipedeSubstrateChange> {
  const { data } = await apiClient.post<CentipedeSubstrateChange>(
    `/centipedes/${centipedeId}/substrate-changes`,
    payload,
  );
  return data;
}

export async function listCentipedePhotos(
  centipedeId: string,
): Promise<CentipedePhoto[]> {
  const { data } = await apiClient.get<CentipedePhoto[]>(
    `/centipedes/${centipedeId}/photos`,
  );
  return data;
}

/** Multipart photo upload — caller passes a FormData (image file + optional caption). */
export async function uploadCentipedePhoto(
  centipedeId: string,
  form: FormData,
): Promise<CentipedePhoto> {
  const { data } = await apiClient.post<CentipedePhoto>(
    `/centipedes/${centipedeId}/photos`,
    form,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return data;
}

// ---------------------------------------------------------------------------
// QR upload session
// ---------------------------------------------------------------------------

export interface CentipedeUploadSession {
  token: string;
  upload_url: string;
  expires_at: string;
  expires_in_minutes: number;
  taxon: 'centipede';
  centipede_name: string;
}

export async function createCentipedeUploadSession(
  centipedeId: string,
): Promise<CentipedeUploadSession> {
  const { data } = await apiClient.post<CentipedeUploadSession>(
    `/centipedes/${centipedeId}/upload-session`,
  );
  return data;
}

// ---------------------------------------------------------------------------
// Display helpers
// ---------------------------------------------------------------------------

/** "Stinger (Scolopendra heros)" → tries name first, falls back to species. */
export function centipedeDisplayName(c: Centipede): string {
  const parts: string[] = [];
  if (c.name) parts.push(c.name);
  const label = c.common_name || c.scientific_name;
  if (label) parts.push(c.name ? `(${label})` : label);
  return parts.join(' ') || 'Unnamed centipede';
}

/** Venom-tier pill colors — same palette as scorpions so the collection
 * card can show a consistent "hot keepers only" badge across taxa. */
export function venomSeverityColor(
  severity: VenomSeverity | null | undefined,
): { bg: string; fg: string; label: string } {
  switch (severity) {
    case 'medically_significant':
      return { bg: '#fee2e2', fg: '#991b1b', label: 'Hot' };
    case 'moderate':
      return { bg: '#fef3c7', fg: '#92400e', label: 'Moderate' };
    case 'mild':
      return { bg: '#dcfce7', fg: '#166534', label: 'Mild' };
    default:
      return { bg: '#e5e7eb', fg: '#374151', label: 'Unknown' };
  }
}

File: apps/api/app/routers/centipedes.py
"""Centipede routes — Phase C2 of ADR-005.

  GET    /api/v1/centipedes/                 list current user's centipedes
  POST   /api/v1/centipedes/                  create
  GET    /api/v1/centipedes/{centipede_id}    fetch one
  PUT    /api/v1/centipedes/{centipede_id}    partial update
  DELETE /api/v1/centipedes/{centipede_id}    delete (cascades to logs)

Centipedes live exclusively on the unified `inverts` table with
`taxon='centipede'`. There is NO legacy `centipedes` table — they
launched directly on the consolidated surface, so there's no
dual-write to maintain.

This router is structurally identical to the scorpion + tarantula
routers (keeper-UX consistency), but its data plane is `inverts WHERE
taxon='centipede'`. After Phase D ships and the legacy tarantula +
scorpion tables drop, all three per-taxon routers will share this
shape.

Ownership: every query filters by Invert.user_id == current_user.id
AND Invert.taxon == 'centipede'. A tarantula ID can't be retrieved or
modified via this router — wrong taxon = 404.
"""
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.invert import Invert
from app.models.invert_species import InvertSpecies
from app.models.tarantula import Sex, Source  # shared DB enums (UPPERCASE)
from app.models.user import User
from app.schemas.centipede import (
    CentipedeCreate, CentipedeResponse, CentipedeUpdate,
)
from app.utils.dependencies import get_current_user
from app.utils.limits import enforce_collection_limit

router = APIRouter()


TAXON = "centipede"


def _coerce_enums(data: dict) -> dict:
    """Map string sex / source onto the shared DB enums (UPPERCASE
    names in prod). Same pattern as tarantula + scorpion + invert routers."""
    if data.get("sex"):
        try:
            data["sex"] = Sex(data["sex"])
        except ValueError:
            pass
    if data.get("source"):
        try:
            data["source"] = Source(data["source"])
        except ValueError:
            pass
    return data


def _validate_species(db: Session, species_id: Optional[UUID]) -> None:
    """Hard 400 if a non-centipede species is attached. Prevents a
    keeper from linking a centipede record to a tarantula or scorpion
    care sheet."""
    if species_id is None:
        return
    species = db.query(InvertSpecies).filter(
        InvertSpecies.id == species_id,
    ).first()
    if species is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Species not found",
        )
    if species.taxon != TAXON:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Species belongs to taxon '{species.taxon}', not 'centipede'",
        )


def _owned_centipede(
    db: Session, centipede_id: UUID, user: User,
) -> Invert:
    """Lookup-or-404 with both ownership AND taxon checks. Returning the
    row from one helper means every CRUD handler shares the same gate."""
    row = db.query(Invert).filter(
        Invert.id == centipede_id,
        Invert.user_id == user.id,
        Invert.taxon == TAXON,
    ).first()
    if not row:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Centipede not found",
        )
    return row


@router.get("/", response_model=List[CentipedeResponse])
async def list_centipedes(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """List the authenticated user's centipedes, newest first."""
    return (
        db.query(Invert)
        .filter(
            Invert.user_id == current_user.id,
            Invert.taxon == TAXON,
        )
        .order_by(Invert.created_at.desc())
        .all()
    )


@router.post(
    "/", response_model=CentipedeResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_centipede(
    payload: CentipedeCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a new centipede. Taxon is force-set to 'centipede'."""
    # Cross-taxon collection cap (counts inverts: tarantulas + scorpions + centipedes).
    enforce_collection_limit(db, current_user)
    _validate_species(db, payload.species_id)

    data = _coerce_enums(payload.model_dump())

    # Default visibility to the owner's profile preference, matching
    # the tarantula + scorpion path.
    if not data.get("visibility"):
        data["visibility"] = (
            "public" if current_user.collection_visibility == "public"
            else "private"
        )

    new_centipede = Invert(
        user_id=current_user.id,
        taxon=TAXON,
        **data,
    )
    db.add(new_centipede)
    db.commit()
    db.refresh(new_centipede)

    if new_centipede.species_id:
        species = db.query(InvertSpecies).filter(
            InvertSpecies.id == new_centipede.species_id,
        ).first()
        if species:
            species.times_kept = (species.times_kept or 0) + 1
            db.commit()

    return new_centipede


@router.get(
    "/{centipede_id}", response_model=CentipedeResponse,
)
async def get_centipede(
    centipede_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Fetch one. 404 if not found OR if the row's taxon isn't 'centipede'."""
    return _owned_centipede(db, centipede_id, current_user)


@router.put(
    "/{centipede_id}", response_model=CentipedeResponse,
)
async def update_centipede(
    centipede_id: UUID,
    payload: CentipedeUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Partial update — only fields present in the payload are applied.
    Taxon is omitted from the schema; it can't change."""
    centipede = _owned_centipede(db, centipede_id, current_user)

    data = payload.model_dump(exclude_unset=True)

    if "species_id" in data:
        _validate_species(db, data["species_id"])

    data = _coerce_enums(data)
    for field, value in data.items():
        setattr(centipede, field, value)

    db.commit()
    db.refresh(centipede)
    return centipede


@router.delete(
    "/{centipede_id}", status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_centipede(
    centipede_id: UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Delete. Cascades to feeding/molt/substrate logs, photos, and QR
    sessions via the FK ON DELETE CASCADE clauses on `invert_id`."""
    centipede = _owned_centipede(db, centipede_id, current_user)
    db.delete(centipede)
    db.commit()
    return None
